policy/predicate/signature.go

```diff
@@ -98,29 +136,53 @@ type HasValidSignaturesByKeys struct {
 
 var _ Predicate = &HasValidSignaturesByKeys{}
 
-func (pred *HasValidSignaturesByKeys) Evaluate(ctx context.Context, prctx pull.Context) (bool, string, error) {
+func (pred *HasValidSignaturesByKeys) Evaluate(ctx context.Context, prctx pull.Context) (*common.PredicateResult, error) {
 	commits, err := prctx.Commits()
+
+	predicateResult := common.PredicateResult{
+		Satisfied:       false,
+		ConditionPhrase: "have valid signatures keys",
+		ConditionValues: pred.KeyIDs,
+	}
+
 	if err != nil {
-		return false, "", errors.Wrap(err, "failed to get commits")
+		return &predicateResult, errors.Wrap(err, "failed to get commits")
 	}
 
-	keys := make(map[string]struct{})
+	keys := make(map[string][]string)
+
+	var commitHashes []string
 
 	for _, c := range commits {
 		valid, desc := hasValidSignature(ctx, c)
 		if !valid {
-			return false, desc, nil
+			predicateResult.Values = []string{c.SHA}
+			predicateResult.Description = desc
+			predicateResult.ValuePhrase = "commit"
+			return &predicateResult, nil
 		}
+		commitHashes = append(commitHashes, c.SHA)
 		// Only GPG signatures are valid for this predicate
 		switch c.Signature.Type {
 		case pull.SignatureGpg:
-			keys[c.Signature.KeyID] = struct{}{}
+			commitSHAs, ok := keys[c.Signature.KeyID]
+			if ok {
+				keys[c.Signature.KeyID] = append(commitSHAs, c.SHA)
+			} else {
+				keys[c.Signature.KeyID] = []string{c.SHA}
+			}
 		default:
-			return false, fmt.Sprintf("Commit %.10s signature is not a GPG signature", c.SHA), nil
+			predicateResult.Values = []string{"not GPG"}
+			predicateResult.ValuePhrase = "signature type"
```

Comment:
```suggestion
			predicateResult.Values = []string{c.SHA}
			predicateResult.ValuePhrase = "commits" // this might already be set
			predicateResult.ConditionPhrase = "have GPG signatures"
```